fedimint-api/src/db/mod.rs

```diff
@@ -56,16 +57,83 @@ pub trait DatabaseValue: Sized + SerializableDatabaseValue {
 }
 
 pub type PrefixIter<'a> = Box<dyn Iterator<Item = Result<(Vec<u8>, Vec<u8>)>> + Send + 'a>;
+type Notifications = DashMap<Box<[u8]>, Arc<Notify>>;
 
 #[async_trait]
 pub trait IDatabase: Debug + Send + Sync {
-    async fn begin_transaction(&self, decoders: ModuleDecoderRegistry) -> DatabaseTransaction;
+    async fn begin_transaction<'a>(&'a self) -> Box<dyn IDatabaseTransaction<'a> + Send + 'a>;
 }
 
-dyn_newtype_define! {
-    /// A handle to a type-erased database implementation
-    #[derive(Clone)]
-    pub Database(Arc<IDatabase>)
+#[derive(Clone, Debug)]
+pub struct Database {
+    inner: Arc<dyn IDatabase>,
+    notifications: Arc<Notifications>,
+}
+
+impl<D: IDatabase + 'static> From<D> for Database {
+    fn from(inner: D) -> Self {
+        Database {
+            inner: Arc::new(inner),
+            notifications: Arc::new(Notifications::new()),
+        }
+    }
+}
+
+impl Database {
+    pub async fn begin_transaction(&self, decoders: ModuleDecoderRegistry) -> DatabaseTransaction {
+        DatabaseTransaction::new(
+            self.inner.begin_transaction().await,
+            decoders,
+            self.notifications.clone(),
+        )
+    }
+
+    pub async fn wait_key<K: DatabaseKey + DatabaseKeyPrefixConst>(
+        &self,
+        key: K,
+        decoders: ModuleDecoderRegistry,
+    ) -> Result<K::Value> {
+        let mut tx = self.inner.begin_transaction().await;
+        let key_bytes = key.to_bytes();
+        let value_bytes = if let Some(value_bytes) = tx.raw_get_bytes(&key.to_bytes()).await? {
+            value_bytes
+        } else {
+            struct CancelGaurd<F: FnOnce()>(Option<F>);
+            impl<F: FnOnce()> Drop for CancelGaurd<F> {
+                fn drop(&mut self) {
+                    self.0.take().unwrap()();
+                }
+            }
+            let notifications = self.notifications.clone();
+            // maybe use Arc instead?
+            let key_bytes2 = key_bytes.clone();
+            // cancellation behaviour: if one wait_key gets cancelled, all others with same key will
+            // never return, this is correct enough in case of timeouts
+            let _gaurd = CancelGaurd(Some(move || {
+                // only removes it first time
+                notifications.remove(key_bytes2.as_slice());
+            }));
+            let notif: Arc<Notify> = self
```

Comment:
Is anything ever removing elements from `self.notifications` which will lead to it blowing up over time, no?

Reply:
Yes, the cancel guard. It also runs even if the future isn't cancelled. Maybe `Deferred` would be a better name.

Reply:
This might not be an easy problem to overcome. We could potentially just have 'per-prefix' instead of per-key `Notify`, to bound the number of `Notify` we can have. In practice we probably don't do that many updates anyway.

Even further, we might have just one shared `Notify` for everyone.

Reply:
It's kind of a time/memory trade off imo. With one `Notify` we will waste a lot of io/compute by making every websocket connection check if it can send an update, with one notify per prefix we need too much memory and need to create `Notify`s too often. Using prefixes would not be a good balance between these two imo since certain prefixes will be subscrived to much more often.

An alternative approach, similar to the prefix approach but more balanced over all `Notify` instances would be hashing the key and then assigning a `Notify` instance based on a hash prefix, much like buckets of a hash map. We can scale the amount of `Notify`s as needed.

Reply:
I am sold on hash based approach. It would also simplify the code: now the notifies are already initialized in a vec with specific size, you don't need to insert or remove it. which fixes the cancellation issues this has (see cancellation comment in code).